### DISCUSSION TITLE
Automatic Mesh service

### DISCUSSION THREAD

**Discussion by ghost** [Ylianst/MeshCentral#6099] [Issues]
Hi,
Sorry for asking again what I'll ask now.
I've reinstalled the mesh and again I have an issue with meshcentral.service

The Service script:

[Unit]
Description=MeshCentral Server
After=network.target

[Service]
Type=simple

User=quizhead
Group=quizhead

WorkingDirectory=/home/quizhead/meshcentral/node_modules/meshcentral
ExecStart=/usr/bin/node meshcentral

[Install]
WantedBy=multi-user.target 

I'm getting the error:
`root@Latitude5480:~# sudo systemctl status meshcentral 
● meshcentral.service - MeshCentral Server
     Loaded: loaded (/etc/systemd/system/meshcentral.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2024-05-16 18:57:10 CEST; 5s ago
   Main PID: 4415 (node)
      Tasks: 11 (limit: 18888)
     Memory: 18.0M
        CPU: 2.053s
     CGroup: /system.slice/meshcentral.service
             └─4415 /usr/bin/node meshcentral

May 16 18:57:11 Latitude5480 node[4415]:     at ChildProcess.exithandler (child_process.js:308:12)
May 16 18:57:11 Latitude5480 node[4415]:     at ChildProcess.emit (events.js:326:22)
May 16 18:57:11 Latitude5480 node[4415]:     at maybeClose (internal/child_process.js:1022:16)
May 16 18:57:11 Latitude5480 node[4415]:     at Process.ChildProcess._handle.onexit (internal/child_process.js:287:5) {
May 16 18:57:11 Latitude5480 node[4415]:   killed: false,
May 16 18:57:11 Latitude5480 node[4415]:   code: 1,
May 16 18:57:11 Latitude5480 node[4415]:   signal: null,
May 16 18:57:11 Latitude5480 node[4415]:   cmd: '/usr/bin/node /home/quizhead/meshcentral/node_modules/meshcentral/meshcentral --launch 4415'
May 16 18:57:11 Latitude5480 node[4415]: }
May 16 18:57:11 Latitude5480 node[4415]: ERROR: MeshCentral failed with critical error, check mesherrors.txt. Restarting in 5 seconds...`

Is the messerror I see:
`/home/quizhead/meshcentral/node_modules/meshcentral/node_modules/minimatch/dist/commonjs/ast.js:86
    #fillNegs() {
             ^

SyntaxError: Unexpected token '('
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/home/quizhead/meshcentral/node_modules/meshcentral/node_modules/minimatch/dist/commonjs/index.js:9:18)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)`

Node -v  gives v20.13.1

```root@Latitude5480:/home/quizhead# ls -lah /home/quizhead/meshcentral/meshcentral-data/
total 7.2M
drwxrwxr-x 3 quizhead quizhead 4.0K May 16 19:00 .
drwxrwxr-x 6 quizhead quizhead 4.0K May 16 12:34 ..
-rw-rw-r-- 1 quizhead quizhead 2.5K May 16 08:56 agentserver-cert-private.key
-rw-rw-r-- 1 quizhead quizhead 1.6K May 16 08:56 agentserver-cert-public.crt
-rw-rw-r-- 1 quizhead quizhead 2.5K May 16 08:56 codesign-cert-private.key
-rw-rw-r-- 1 quizhead quizhead 1.6K May 16 08:56 codesign-cert-public.crt
-rw-rw-r-- 1 quizhead quizhead  764 May 16 12:14 config.json
-rw-r--r-- 1 quizhead quizhead  44K May 16 19:00 meshcentral.db
-rw-r--r-- 1 quizhead quizhead 116K May 16 19:00 meshcentral-events.db
-rw-r--r-- 1 quizhead quizhead 221K May 16 19:00 meshcentral-power.db
-rw-r--r-- 1 quizhead quizhead  40K May 16 19:00 meshcentral-stats.db
-rw-r--r-- 1 quizhead quizhead 2.0M May 16 19:00 mesherrors.txt
-rw-r--r-- 1 quizhead quizhead 1.7K May 16 14:22 mpsserver-cert-private.key
-rw-rw-r-- 1 quizhead quizhead 1.7K May 16 08:56 mpsserver-cert-private.key-old
-rw-r--r-- 1 quizhead quizhead 1.5K May 16 14:22 mpsserver-cert-public.crt
-rw-rw-r-- 1 quizhead quizhead 1.5K May 16 08:56 mpsserver-cert-public.crt-old
-rw-rw-r-- 1 quizhead quizhead 2.5K May 16 08:56 root-cert-private.key
-rw-rw-r-- 1 quizhead quizhead 1.6K May 16 08:56 root-cert-public.crt
-rw-rw-r-- 1 quizhead quizhead   95 May 16 19:00 serverstate.txt
drwxrwxr-x 2 quizhead quizhead 4.0K May 16 08:56 signedagents
-rw-r--r-- 1 quizhead quizhead 2.5K May 16 14:22 webserver-cert-private.key
-rw-rw-r-- 1 quizhead quizhead 2.5K May 16 08:56 webserver-cert-private.key-old
-rw-r--r-- 1 quizhead quizhead 1.7K May 16 14:22 webserver-cert-public.crt
-rw-rw-r-- 1 quizhead quizhead 1.7K May 16 08:56 webserver-cert-public.crt-old
-rw-r--r-- 1 quizhead quizhead 4.7M May 16 13:43 WelcomePic.JPG```

I've made the translation process as well

Thanks again.

**Comment by si458** [comment]
It's complaining of an old node version
Check by running `/usr/bin/node -v` and see what node version gets returned
Then check where `node -v` is located by doing `which node` 
Then the output from `which node` u need to edit `/etc/systemd/system/meshcentral.service` with the correct path

**Comment by ghost** [comment]
which node gives : /root/.nvm/versions/node/v20.13.1/bin/node
The service gives:
× meshcentral.service - MeshCentral Server
     Loaded: loaded (/etc/systemd/system/meshcentral.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Thu 2024-05-16 21:13:40 CEST; 2min 53s ago
   Main PID: 20898 (code=exited, status=203/EXEC)
        CPU: 2ms

May 16 21:13:40 Latitude5480 systemd[1]: Started MeshCentral Server.
May 16 21:13:40 Latitude5480 systemd[20898]: meshcentral.service: Failed to execute /root/.nvm/versions/node/v20.13.1/bin/node: Permission denied
May 16 21:13:40 Latitude5480 systemd[20898]: meshcentral.service: Failed at step EXEC spawning /root/.nvm/versions/node/v20.13.1/bin/node: Permission denied
May 16 21:13:40 Latitude5480 systemd[1]: meshcentral.service: Main process exited, code=exited, status=203/EXEC
May 16 21:13:40 Latitude5480 systemd[1]: meshcentral.service: Failed with result 'exit-code'.

This mean that node is installed on the root and not the user if I understand correctly.
How can I fix it?
Thanks.

**Reply by si458** [reply]
Make sure the execstart is `nodepath meshcentral` and also u have the service running as one user but node was installed into root.
You need to login as the meshcentral user u want then install nvm there and check its path

**Reply by si458** [reply]
Don't use sudo with nvm!!!

**Reply by ghost** [reply]
I did nvm install node without sudo but still I see v12.22.9 when I run /usr/bin/node -v
The which node command also gives the same /root/.nvm/versions/node/v22.2.0/bin/node
I also tried nodesource install but nodesource: command not found

**Reply by si458** [reply]
So make sure u have removed the old node as u might have installed it via apt
`apt remove node` `apt remove nodejs`
The is no such command for nodesource as nodesource is just a repo

**Comment by ghost** [comment]
This node thing is confusing me.
I removed the node, restarted the machine and when I run nvm install node it says that it's already installed
When I run /usr/bin/node -v it says that there is no file or directory
Then when I run which node from the user I see /quizhead/.nvm/versions/node/v22.2.0/bin/node instead of root.
When I try to run the mesh service I get:
```× meshcentral.service - MeshCentral Server
     Loaded: loaded (/etc/systemd/system/meshcentral.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Thu 2024-05-16 21:48:18 CEST; 17min ago
   Main PID: 826 (code=exited, status=203/EXEC)
        CPU: 2ms
May 16 21:48:18 Latitude5480 systemd[1]: Started MeshCentral Server.
May 16 21:48:18 Latitude5480 systemd[1]: meshcentral.service: Main process exited, code=exited, status=203/EXEC
May 16 21:48:18 Latitude5480 systemd[1]: meshcentral.service: Failed with result 'exit-code'.```

I'm not sure what went wrong here

**Reply by si458** [reply]
OK so now it seems correct,
Ur service file is incorrect still
It should look the following here in the docs https://ylianst.github.io/MeshCentral/install/install2/#automatically-starting-the-server_2
Except u change `/usr/bin/node` to the path of `which node`

**Reply by ghost** [reply]
I did but still the same error
```
[Unit]
Description=MeshCentral Server

[Service]
Type=simple
LimitNOFILE=1000000
ExecStart=/quizhead/.nvm/versions/node/v22.2.0/bin/node
WorkingDirectory=/home/quizhead/meshcentral/node_modules/meshcentral
Environment=NODE_ENV=production
User=default
Group=default
Restart=always
# Restart service after 10 seconds if node service crashes
RestartSec=10
# Set port permissions capability
AmbientCapabilities=cap_net_bind_service

[Install]
WantedBy=multi-user.target
```

**Reply by si458** [reply]
Ur execstart and working directory aren't the same as in the docs if u look
They should be the following
```
ExecStart=/quizhead/.nvm/versions/node/v22.2.0/bin/node /home/quizhead/meshcentral/node_modules/meshcentral
WorkingDirectory=/home/quizhead/meshcentral
```

**Reply by ghost** [reply]
Still doesn't work after udpating the service

```● meshcentral.service - MeshCentral Server
     Loaded: loaded (/etc/systemd/system/meshcentral.service; enabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Thu 2024-05-16 23:28:55 CEST; 2s ago
    Process: 9151 ExecStart=/quizhead/.nvm/versions/node/v22.2.0/bin/node /home/quizhead/meshcentral/node_modules/meshcentral (code=exited, status=217/USER)
   Main PID: 9151 (code=exited, status=217/USER)
        CPU: 2ms```

**Reply by si458** [reply]
OK so what output do you get if u 
`cd /home/quizhead/meshcentral`
Then
`/quizhead/.nvm/versions/node/v22.2.0/bin/node /home/quizhead/meshcentral/node_modules/meshcentral` ?

**Reply by ghost** [reply]
bash: /quizhead/.nvm/versions/node/v22.2.0/bin/node: No such file or directory
But I see the path when I navigate manually

**Reply by si458** [reply]
Are u sure it's `/quizhead/` and not `/home/quizhead` ur home directory?

As u say above ur working directory is `/home/quizhead/meshcentral` so in theory

The ExecStart would be `/home/quizhead/.nvm/versions/node/v22.2.0/bin/node /home/quizhead/meshcentral/node_modules/meshcentral`

Sorry too many messages/mixed paths etc all over the place

**Reply by ghost** [reply]
I tried it already but I'm going to reinstall the whole system...again.... as this became too complicated.
I am following the docs and doing exactly as written so I don't know what goes wrong each time I'm doing it.
It's only this Node shit that gets corrupted each time I install the mesh which affects the mesh service.
Anyway you can delete this post and I will try again from scratch.
As allways, thank you for the help.

**Reply by ghost** [reply]
OK  i've removed the Node completely from local user and root side and the Mesh
Reinstalled everything only from the local user side, used the which node to update the service and now it works.
I must have by mistake installed on both locations and everything got mixed.
Took this opportunity to see if the backup as well works and it does.
Sorry for the trouble and thanks for the help.
